sklearn/impute/_base.py

```diff
@@ -408,17 +412,23 @@ def transform(self, X):
         check_is_fitted(self)
 
         X = self._validate_input(X)
-        X_indicator = super()._transform_indicator(X)
-
         statistics = self.statistics_
 
         if X.shape[1] != statistics.shape[0]:
             raise ValueError("X has %d features per sample, expected %d"
                              % (X.shape[1], self.statistics_.shape[0]))
 
+        # compute mask and missing_mask before eliminating invalid features
+        if sparse.issparse(X):
+            mask, missing_mask = _get_mask(X, self.missing_values,
+                                           reconstruct_sparse=True)
+        else:
+            mask = _get_mask(X, self.missing_values)
+
         # Delete the invalid columns if strategy is not constant
         if self.strategy == "constant":
             valid_statistics = statistics
+            valid_statistics_indexes = slice(None)
```

Comment:
This is not straigtforward here. Why could we not use `valid_statistics_indexes = None`?

Reply:
That would create a problem here -

https://github.com/harsh020/scikit-learn/blob/2ee53e287884bf1727b746ab7fb0c91b799f944d/sklearn/impute/_base.py#L463-L466

If the `valid_statistical_indexes` is `None`, then the matrix of `X` would be converted to a tensor where each old-row will become a matrix.

Reply:
For maintenance purposes, I do agree that `slice(None)` is not too clear, especially when we use it later like this

```py
if not isinstance(valid_statistics_indexes, slice):
```

Lets go with `None` and then with the `mask_valid_features`:
```py

if valid_statistics_indexes is None:
	mask_valid_features = missing_mask
else:
	mask_valid_features = missing_mask[:, valid_statistics_indexes]
```